README
NAME

    MooseX::Util - Moose::Util extensions

VERSION

    This document describes version 0.005 of MooseX::Util - released May
    15, 2015 as part of MooseX-Util.

SYNOPSIS

        use MooseX::Util qw{ ensure_all_roles with_traits };
    
        # profit!

DESCRIPTION

    This is a utility module that handles all of the same functions that
    Moose::Util handles. In fact, most of the functions exported by this
    package are simply re-exports from Moose::Util, so you're recommended
    to read the documentation of that module for a comprehensive view.

    However.

    We've re-implemented a number of the functions our parent provides, for
    a variety of reasons. Those functions are documented here.

FUNCTIONS

 with_traits(<classname> => (<trait1>, ... ))

    Given a class and one or more traits, we construct an anonymous class
    that is a subclass of the given class and consumes the traits given.
    This is exactly the same as "with_traits" in Moose::Util, except that
    we use "create_anon_class" in MooseX::Util::Meta::Class to construct
    the anonymous class, rather than "create_anon_class" in
    Moose::Meta::Class directly.

    Essentially, this means that when we do:

        my $anon_class_name = with_traits('Zombie::Catcher', 'SomeTrait');

    For $anon_class_name we get:

        Zombie::Catcher::__ANON__::SERIAL::1

    Rather than:

        Moose::Meta::Class::__ANON__::SERIAL::1

    This is nice because we have an idea of where the first anonymous class
    came from, whereas the second one could could be from anywhere.

 is_private

        # true if "private"
        ... if is_private('_some_name');

    Ofttimes we need to determine if a name is considered "private" or not.
    By convention, method, attribute, and other names are considered
    private if their first character is an underscore.

    While trivial to test for, this allows us to centralize the tests in
    one place.

SEE ALSO

    Please see those modules/websites for more information related to this
    module.

      * Moose::Util

BUGS

    Please report any bugs or feature requests on the bugtracker website
    https://github.com/RsrchBoy/moosex-util/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR

    Chris Weyl <cweyl@alumni.drew.edu>

 I'm a material boy in a material world

    Please note I do not expect to be gittip'ed or flattr'ed for this work,
    rather it is simply a very pleasant surprise. I largely create and
    release works like this because I need them or I find it enjoyable;
    however, don't let that stop you if you feel like it ;)

    Flattr
    <https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithub.com%2FRsrchBoy%2Fmoosex-util&title=RsrchBoy's%20CPAN%20MooseX-Util&tags=%22RsrchBoy's%20MooseX-Util%20in%20the%20CPAN%22>,
    Gratipay <https://gratipay.com/RsrchBoy/>, or indulge my Amazon
    Wishlist <http://bit.ly/rsrchboys-wishlist>... If and *only* if you so
    desire.

COPYRIGHT AND LICENSE

    This software is Copyright (c) 2012 by Chris Weyl.

    This is free software, licensed under:

      The GNU Lesser General Public License, Version 2.1, February 1999



Changes
Revision history for MooseX-Util

{{$NEXT}}

0.006     2015-06-26 11:32:06 PDT-0700
	* Drop MXAS, as we don't need it.

0.005     2015-05-15 09:38:31 PDT-0700
	* add a stub find_meta...  ...as apparently a number of items calling this
	  as MooseX::Util::find_meta() were breaking after the 0.004 changes.

0.004     2015-05-07 21:58:06 PDT-0700
	* add: is_private()

0.003     2014-05-22 12:30:56 PDT-0700
	* MooseX::TraitFor::Meta::Class::BetterAnonClassNames has been split into
	  its own dist, so let's require the first version of the module present
	  in that dist (and drop it here, naturally).

0.002     2014-03-03 12:17:31 PST-0800
	* Drop Reindeer as a prereq (oops)
	* Add a cpanfile
	* Mark v5.10 as the, ah, maximum minimum version of Perl

0.001     2012-10-29 11:20:28 PST8PDT
	* Initial release
	* we only override with_traits() at this point (in MX::Utils), in order to
	  use our slightly nicer set of anonymous role name picking
	* also includes MooseX::Traitor, and the various meta-bits needed to do
	  the above.  Traitor aims to be a MX::Traits alternate, but right now
	  it's pretty limited.


Makefile.PL
#
# This file is part of MooseX-Util
#
# This software is Copyright (c) 2012 by Chris Weyl.
#
# This is free software, licensed under:
#
#   The GNU Lesser General Public License, Version 2.1, February 1999
#
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.036.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Moose::Util extensions",
  "AUTHOR" => "Chris Weyl <cweyl\@alumni.drew.edu>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "MooseX-Util",
  "EXE_FILES" => [],
  "LICENSE" => "lgpl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "MooseX::Util",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Moose" => 0,
    "Moose::Meta::Class" => 0,
    "Moose::Role" => 0,
    "Moose::Util" => 0,
    "MooseX::AttributeShortcuts" => 0,
    "MooseX::TraitFor::Meta::Class::BetterAnonClassNames" => "0.002001",
    "Sub::Exporter::Progressive" => 0,
    "namespace::autoclean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::Moose::More" => "0.016",
    "Test::More" => "0.94",
    "Test::Requires" => 0,
    "aliased" => 0
  },
  "VERSION" => "0.005",
  "test" => {
    "TESTS" => "t/*.t t/metaclass/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moose" => 0,
  "Moose::Meta::Class" => 0,
  "Moose::Role" => 0,
  "Moose::Util" => 0,
  "MooseX::AttributeShortcuts" => 0,
  "MooseX::TraitFor::Meta::Class::BetterAnonClassNames" => "0.002001",
  "Sub::Exporter::Progressive" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::Moose::More" => "0.016",
  "Test::More" => "0.94",
  "Test::Requires" => 0,
  "aliased" => 0,
  "namespace::autoclean" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/MooseX/Traitor.pm
package MooseX::Traitor;

# ABSTRACT: An alternate way to compose your classes with traits

use Moose::Role;
use namespace::autoclean;
use MooseX::Util ();

=method with_traits(<trait1>, ...)

This method builds an anonymous class from the consuming class and any traits
specified.

You may use the full trait specification syntax, e.g.:

    MyClass->with_traits('My::Trait' => { -excludes => ... })

Calling this routine with no traits specified will simply return the name of
the class.  This is not considered an error.

Note that we handle being called directly against a package (e.g.
C<< MyClass->with_traits(...) >>) and against an instance (e.g.
C<< $self->with_traits(...) >>) identically; in each instance the class
referenced is subclassed.

=cut

sub with_traits {
    my ($thing, @traits) = @_;

    my $class = blessed $thing || $thing;

    return MooseX::Util::with_traits($class => @traits);
}

!!42;
__END__

=for stopwords composable CLOS behaviour behaviours

=head1 SYNOPSIS

    # in your class definition...
    package MyClass;
    use Moose;
    use namespace::autoclean;

    with 'MooseX::Traitor';

    # somewhere else in Gotham...
    my $thinger = MyClass->with_traits('Thinger::Trait1')->new(...);

=head1 DESCRIPTION

One of the most powerful things about L<Moose> is that with roles and easy
"anonymous" class creation we are blessed with a fantastic new way of
creating classes, often on the fly, out of other classes and those composable
bits of behaviour, roles.

Even better, this application of discrete chunks of behaviours enables people
simply using a class to extend and tweak its behaviour in new ways -- possibly
ways never contemplated by the authors of the classes being altered.

=head1 ROLES OR TRAITS?

There are many different definitions of what a role is vs a trait, ranging
from "hey man, it's all cool" to "CLOS calls them all traits SO TRAITS IS THE
ONE TRUE NAME", it seems that most people tend to think of them this way:

Roles are traits that a class knowingly consumes (e.g. via with()).

Traits are roles that are applied without the class' consent (e.g. anonymous
subclass composition or C<< $trait_meta->apply('ClassThinger') >>).

Or maybe that's just what this author is imposing on everyone else.  Either
way, that's what we'll be using here if the definition ever becomes important.

=cut


lib/MooseX/Util/Meta/Class.pm
package MooseX::Util::Meta::Class;

# ABSTRACT: A helper metaclass

use Moose;
use namespace::autoclean;

extends 'Moose::Meta::Class';
with 'MooseX::TraitFor::Meta::Class::BetterAnonClassNames';

# NOTE: making this package immutable breaks our metaclass compatibility!
#__PACKAGE__->meta->make_immutable;
!!42;
__END__

=for :stopwords BetterAnonClassNames

=head1 SYNOPSIS

    # create a new type of Zombie catcher equipped with machete and car
    my $meta = MooseX::Util::Meta::Class->create_anon_class(
        'Zombie::Catcher' => qw{
            Zombie::Catcher::Tools::Machete
            Zombie::Catcher::Tools::TracyChapmansFastCar
         },
     );

    # created anon classname is: Zombie::Catcher::__ANON__::SERIAL::42

=head1 DESCRIPTION

This is a trivial extension of L<Moose::Meta::Class> that consumes the
L<BetterAnonClassNames|MooseX::TraitFor::Meta::Class::BetterAnonClassNames>
trait.

=cut


lib/MooseX/TraitFor/Moose/Meta/Class/BetterAnonClassNames.pm
package MooseX::TraitFor::Moose::Meta::Class::BetterAnonClassNames;

# ABSTRACT: Metaclass trait

use Moose::Role;
use namespace::autoclean;
use MooseX::AttributeShortcuts;
use autobox::Core;

# debugging...
#use Smart::Comments '###';

has is_anon => (is => 'ro', isa => 'Bool', default => 0);

has anon_package_prefix => (
    is       => 'ro',
    isa      => 'Str',
    builder  => 1,
    #init_arg => 'anon_package_prefix',
);

sub _build_anon_package_prefix { Moose::Meta::Class->_anon_package_prefix }

#sub _anon_package_prefix { shift->name . '::Package::__ANON__::SERIAL::' }
#has _anon_package_prefix => (
    #is      => 'rwp',
    #isa     => 'Str',
    #clearer => 1,
    #builder => sub { 'Reindeer' . shift->_anon_package_middle },
#);

#sub _anon_package_middle { '::Package::__ANON__::SERIAL::' }
sub _anon_package_middle { '::__ANON__::SERIAL::' }

# XXX around?
sub _anon_package_prefix {
    my $thing = shift @_;

    my $prefix = blessed $thing
        ? $thing->anon_package_prefix
        : Moose::Meta::Class->_anon_package_prefix
        ;

    my @caller = caller 1;
    ### @caller
    ### $prefix
    return $prefix;
}

around create => sub {
    my ($orig, $self) = (shift, shift);
    my @args = @_;

    unshift @args, 'package'
        if @args % 2;
    my %opts = @args;

    return $self->$orig(%opts)
        unless $opts{is_anon} && $opts{anon_package_prefix};

    ### old anon package name: $opts{package}
    my $serial = $opts{package}->split(qr/::/)->[-1]; #at(-1); #tail(1);
    $opts{package} = $opts{anon_package_prefix} . $serial;

    ### new anon package name: $opts{package}
    ### %opts
    return $self->$orig(%opts);
};

around create_anon_class => sub {
    my ($orig, $class) = (shift, shift);
    my %opts = @_;

    # we're going to need some additional logic here to make sure that our
    # prefixes make sense; e.g. if we're an anon descendent of an anon class,
    # we should just use our parent's prefix.

    $opts{is_anon} = 1;

    my $superclasses = $opts{superclasses} // [];

    # don't bother doing anything else if we don't have anything to add
    return $class->$orig(%opts)
        if exists $opts{anon_package_prefix};
    return $class->$orig(%opts)
        unless @$superclasses && @$superclasses == 1;

    # XXX ::class_of?
    my $sc = $superclasses->[0]->meta;

    #my $prefix
    $opts{anon_package_prefix}
        = $sc->is_anon
        ? $sc->_anon_package_prefix
        : $superclasses->[0] . $class->_anon_package_middle
        ;

    # basically:  if we have no superclasses, live with the default prefix; if
    # we have a superclass and it's anon, use it's anon prefix; if we have a
    # superclass and it's not anon, make a new prefix out of it
    #
    # cases where we have 2+ superclasses aren't handled right now; we use the
    # Moose::Meta::Class::_anon_package_prefix() for those
    #
    # if we're passed in 'anon_package_prefix', then just use that.

    return $class->$orig(%opts);
};

!!42;


lib/MooseX/Util.pm
#
# This file is part of MooseX-Util
#
# This software is Copyright (c) 2012 by Chris Weyl.
#
# This is free software, licensed under:
#
#   The GNU Lesser General Public License, Version 2.1, February 1999
#
package MooseX::Util;
our $AUTHORITY = 'cpan:RSRCHBOY';
# git description: 0.004-2-ga176eea
$MooseX::Util::VERSION = '0.005';

# ABSTRACT: Moose::Util extensions

use strict;
use warnings;

use parent 'Moose::Util';

use Sub::Exporter::Progressive -setup => {
    exports => [
        qw{
            add_method_modifier
            apply_all_roles
            does_role
            english_list
            ensure_all_roles
            find_meta
            get_all_attribute_values
            get_all_init_args
            is_role
            meta_attribute_alias
            meta_class_alias
            resolve_metaclass_alias
            resolve_metatrait_alias
            search_class_by_role
            throw_exception
            with_traits
        },

        # and our own...
        qw{
            is_private
        },
    ],
    groups => { default => [ ':all' ] },
};

use Carp 'confess';
use MooseX::Util::Meta::Class;


# TODO allow with_traits() to be curried with different class metaclasses?

sub with_traits {
    my ($class, @roles) = @_;
    return $class unless @roles;
    return MooseX::Util::Meta::Class->create_anon_class(
        superclasses => [$class],
        roles        => \@roles,
        cache        => 1,
    )->name;
}


sub is_private($) {
    my ($name) = @_;

    confess 'is_private() must be called with a name to test!'
        unless $name;

    return 1 if $name =~ /^_/;
    return;
}

sub find_meta { goto \&Moose::Util::find_meta }

!!42;

__END__

=pod

=encoding UTF-8

=for :stopwords Chris Weyl

=for :stopwords Wishlist flattr flattr'ed gittip gittip'ed

=head1 NAME

MooseX::Util - Moose::Util extensions

=head1 VERSION

This document describes version 0.005 of MooseX::Util - released May 15, 2015 as part of MooseX-Util.

=head1 SYNOPSIS

    use MooseX::Util qw{ ensure_all_roles with_traits };

    # profit!

=head1 DESCRIPTION

This is a utility module that handles all of the same functions that
L<Moose::Util> handles.  In fact, most of the functions exported by this
package are simply re-exports from L<Moose::Util>, so you're recommended to
read the documentation of that module for a comprehensive view.

However.

We've re-implemented a number of the functions our parent provides, for a
variety of reasons.  Those functions are documented here.

=head1 FUNCTIONS

=head2 with_traits(<classname> => (<trait1>, ... ))

Given a class and one or more traits, we construct an anonymous class that is
a subclass of the given class and consumes the traits given.  This is exactly
the same as L<Moose::Util/with_traits>, except that we use
L<MooseX::Util::Meta::Class/create_anon_class> to construct the anonymous
class, rather than L<Moose::Meta::Class/create_anon_class> directly.

Essentially, this means that when we do:

    my $anon_class_name = with_traits('Zombie::Catcher', 'SomeTrait');

For $anon_class_name we get:

    Zombie::Catcher::__ANON__::SERIAL::1

Rather than:

    Moose::Meta::Class::__ANON__::SERIAL::1

This is nice because we have an idea of where the first anonymous class came
from, whereas the second one could could be from anywhere.

=head2 is_private

    # true if "private"
    ... if is_private('_some_name');

Ofttimes we need to determine if a name is considered "private" or not.  By convention,
method, attribute, and other names are considered private if their first character is
an underscore.

While trivial to test for, this allows us to centralize the tests in one place.

=for Pod::Coverage find_meta

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<Moose::Util>

=back

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/RsrchBoy/moosex-util/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>

=head2 I'm a material boy in a material world

=begin html

<a href="https://gratipay.com/RsrchBoy/"><img src="http://img.shields.io/gratipay/RsrchBoy.svg" /></a>
<a href="http://bit.ly/rsrchboys-wishlist"><img src="http://wps.io/wp-content/uploads/2014/05/amazon_wishlist.resized.png" /></a>
<a href="https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithub.com%2FRsrchBoy%2Fmoosex-util&title=RsrchBoy's%20CPAN%20MooseX-Util&tags=%22RsrchBoy's%20MooseX-Util%20in%20the%20CPAN%22"><img src="http://api.flattr.com/button/flattr-badge-large.png" /></a>

=end html

Please note B<I do not expect to be gittip'ed or flattr'ed for this work>,
rather B<it is simply a very pleasant surprise>. I largely create and release
works like this because I need them or I find it enjoyable; however, don't let
that stop you if you feel like it ;)

L<Flattr|https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithub.com%2FRsrchBoy%2Fmoosex-util&title=RsrchBoy's%20CPAN%20MooseX-Util&tags=%22RsrchBoy's%20MooseX-Util%20in%20the%20CPAN%22>,
L<Gratipay|https://gratipay.com/RsrchBoy/>, or indulge my
L<Amazon Wishlist|http://bit.ly/rsrchboys-wishlist>...  If and *only* if you so desire.

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Chris Weyl.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 2.1, February 1999

=cut
